packages/form-js-viewer/src/features/expressionField/ExpressionLoopPreventer.js

```diff
@@ -0,0 +1,41 @@
+export class ExpressionLoopPreventer {
+  constructor(eventBus) {
+    this._computedExpressions = [];
+
+    eventBus.on('field.updated', ({ shouldNotRecompute }) => {
+      if (shouldNotRecompute) {
+        return;
+      }
+
+      this.reset();
+    });
+
+    eventBus.on('import.done', this.reset.bind(this));
+    eventBus.on('reset', this.reset.bind(this));
```

Comment:
Basically yes, the idea being that we only want expressions to compute once when a change has happened, except if that change was triggered by an expression (in which case we ignore it).

Basically, that makes a change (say a form field value is set) propagate through every expression at least once, but it prevents them from re-evaluating in loops. 

Ideally in the future, we don't have to do any of this sillyness when we separate view and model, but for now it was the best way I found to get around the looping issue entirely within the constraints. It's also why I built it as a module, so that it can be easily extracted.

Reply:
With this system, loops will still loop (they will advance one step per value change). But we are happy having loops be undefined behavior in general, as they are faulty configs. Like you mentioned, linting would be great, but it's hard.

Reply:
Note this does not prevent expressions from chaining, however it does give a certain top to bottom precedence to how they evaluate if the dependency graph has cycles. But again, if they do have cycles it's considered undefined behavior and not something we should care about.

Reply:
That's fine. I am ok with making loops only run once on a change. I do agree that we should separate out the model and the view but that's too big of a change for a fix.